### ISSUE TITLE
MacPass very slow on unlocking

### ISSUE THREAD

**Issue by iuui24** [mstarke/MacPass#550]
On my MBP (2007 Core2Duo running Mavericks) the MacPass.app Version 0.6.2 (12420) is really slow while unlocking the DB (but only there). 

After entering the PW it shows the spinning ball for about 12 sec. before it is ready (just a plain PW, no keyfile etc.).

Other PW apps (eg. KyPassCompanion) need less than 2 sec. with the same DB on the same MBP but are lacking of autotype and others features.

With the search I haven't found anything about this.
Do have other users the same delay? 
Or is it a known issue with a solution? 

Greets,
Ingo

**Comment by mstarke** [comment]
I uses to use some slow high level NSData APIs for data manipulation that might cause this delay. I have to investigate. Since 0.6.2 there are a lot of changes to KeePassKit so this might be fixed. I'll try to publish a new binary ASAP so you can test this agains a newer build.

**Comment by mstarke** [comment]
Is this still valid for the current build (check continuous in the releases)

**Comment by leszekeljasz** [comment]
I was just googling for slowness issue and found this bug report. I'm not sure if this is by design or a bug, but on my MCP 2015 it takes about 1-2s to unlock the DB. I've never experienced it before using various keepass apps on other systems (Windows, Linux, Android). I just tried KeePassX for Mac and it also unlocks instantly.
Just downloaded the most recent build and it behaves exactly the same. Is this expected?

Thanks,
Leszek

**Comment by mstarke** [comment]
Please specify most recent build. (About MacPass -> Version)

This is by no means by design. Could you take a look at the settings for your encryption (AES, TwoFish, …)? Do you use AES as key derivation? What is the number of rounds? Is the database file stored locally or on a network drive? What is the size of the file?

**Comment by leszekeljasz** [comment]
Hi Micheal,

Thanks for replying so fast, details below:
MacPass version: Version 0.7 (20170926160901-continuous)
Encryption: AES Rijandle
Key derivation: AES, 6000 rounds
DB file size: 170KB, I see I have gzip compression enabled in DB settings.
DB file format is kbdx it's stored on local disk (standard Documents dir).

I just created a new DB file in kbdx format and added one entry and MacPass
opens it immediately. Looking at the DB settings and they are nearly the
same, more rounds for Key derivation: 50k.

I'm not sure if 170KB is a lot for keepass DB, but I noticed when searching
with Cmd+F, I had a lot of backup entries, probably created by KeePassX. I
have ~ 200 regular entries in the DB, + ~1200 in backups folder, so total ~
1500 entries. I just deleted the backups, the DB is 88KB now and opens
fast, with very little (but noticeable) delay. The little delay I see now
is perfectly acceptable for me. Thanks for your suggestions, this is much
better now.

Leszek

**Comment by mstarke** [comment]
Thanks for the details. Since you mention a lot of backup entries: Do you use a KDB  a KDBX database? I will try to add a stress test with this amount of entries into my test setups to see if I can speed up performance since this is not acceptable.

**Comment by leszekeljasz** [comment]
I use KBDX. I also noticed that deleting an entry from my big DB takes about half a second, and scrolling stutters. Deleting is much faster after I removed the the backups. None of this is a problem for me, I just figured it might be worth mentioning.

Thanks for great app BTW :-)

**Comment by mstarke** [comment]
Todays continuous build will include some performance enhancements. Please test to see if this makes your experience any better.

**Comment by leszekeljasz** [comment]
What should I be looking for? Opening the DB or manipulating password entries, or...?

I just tried with 20170928140923-continuous and compared it to 20170926160901-continuous (I still had it saved) as well as 0.6.2 (12420). I wasn't able to notice any difference between the two continuous builds when opening the DB file and I tried with two DB sizes: ~1500 entries and ~250. The 0.6.2 build is noticeably slower.

**Comment by mstarke** [comment]
Opening and closing should at least be a bit faster. Manipulating the DB should be noticeable faster.

**Comment by mstarke** [comment]
After having run the profiler for some more, I uncovered that for KDBX3.1 storage a lot of time is used for date parsing. Using KDBX4 (by setting the Key Derivation to Argon2 for instance) yielded 50% faster times for write/read operations. Can you test if this is the case for you as well? I'll try to optimise the date parsing for KDBX3.1.

**Comment by leszekeljasz** [comment]
Changing to Argon2 does make a difference for opening the DB, but manipulating it... I'm not sure, it doesn't feel liek it. I used Argon2 with default settings: 2 rounds, 2 threads and 1M for memory. Are two rounds good enough?

**Comment by mstarke** [comment]
Those settings are the defaults and should be OK. About the improvements upon unlocking, I'm still working on some more speed ups regarding the date parsing.

**Comment by mstarke** [comment]
There are more optimisations present for date parsing in the current continuous build, can you test if this enhances the unlock delay?

**Comment by leszekeljasz** [comment]
What settings should I use? Last time I changed key derivation to Argon and
it's been very fast when unlocking since then. Should I change it back to
AES for testing?

**Comment by mstarke** [comment]
Uh. Change it back to AES before! you use the new version since there was a major bug with KDBX4 storage. After that you can test both. Argon2 should not be any faster than before but AES.

**Comment by mstarke** [comment]
I just tested MacPass 0.6.2 and MacPassHTTP 0.2.1 with chrome and everything worked. Granting access, creating an entry and querying and entry for a web form. I have to check if this is a runtime issue unter 10.8 since I'm on 10.12.

**Comment by mstarke** [comment]
Does 0.7 offer better performance for you?

**Comment by leszekeljasz** [comment]
I'm sorry I missed your previous reply...

I just downloaded 0.7.3 and it somehow corrupted the DB file, or maybe it was a coincidence... I tried to open my DB with 0.7.3 and the app crashed with segmentation fault. I tried multiple times, same result, then I reverted to 0.7.0 and it couldn't open it either - although it was working fine last night. I restored from a backup and 0.7.3 works fine. So not sure what happened there.

0.7 was indeed a bit faster, there is a small lag when opening the DB but it's acceptable. For big database it's slow (still faster than 0.6.2). After deleting the ~1200 backup entries I had in the DB, it's fast enough for me, almost instant. There is other issue that I discovered after using it for some time. I don't want to be picky, it's all good, but since we are in this thread I am just mentioning...
So I'm a shortcut fun, and since I have so many DB entries, I usually search for what I need with Cmf+F. So my typical usage looks like this:
1. Open MacPass
2. Enter DB pass
3. Press Cmd+f to enter search box

But step 3 is not available immediately after the DB unlocks, I still have to wait another 0.5s or so... in this time, pressing Cmd+f only results in this search bar being shown:
<img width="706" alt="screen shot 2017-11-27 at 12 21 17" src="https://user-images.githubusercontent.com/7014321/33264415-c31956f2-d36d-11e7-8f1a-8fcc60f8c63b.png">
but the cursor is not moving to the search box, so when I start typing, nothing happens. I have to press Cmd+f again. However, if I unlock the DB and wait for let's say 1s, and then press Cmd+f, it does open the search bar AND highlights the search box so I can start searching immediately.

So in summary: the DB unlocks fast, I can see the mac pass window displaying a list of entries in the DB, but the UI is not yet fully ready and Cmd+f shortcut does not work consistently.
I know this is very little detail, I'm just a very heavy user and I notice this type of inconsistencies. I did not notice any improvements in this area between 0.7 and 0.7.3. I'mcurrently on 0.7.3.
